env-var-docs/parser-package/src/env_var_docs/parser.py

```diff
@@ -0,0 +1,524 @@
+    """The JSON path of the invalid object."""
+
+    group_errors: ParseErrors
+    """Errors from attempting to parse as a Group."""
+
+    variable_errors: ParseErrors
+    """Errors from attempting to parse as a Variable."""
+
+
+@dataclasses.dataclass
+class Node:
+    """A node within an environment variable documentation file.
+
+    An environment variable documentation file is a JSON object that has
+    object-typed fields. Valid object values are either a Group or Variable.
+    """
+
+    level: int
+    """Level of the node within the file.
+    
+    The environment variable documentation file has an implicit root node, so
+    top-level fields within the file have level 1.
+    """
+
+    json_path: str
+    """Key path of the node. Every path starts with 'file.'"""
+
+    name: str
+    """The group name or variable name. The JSON field key is the name."""
+
+    details: Group | Variable
+    """The group details or variable details."""
+
+
+VisitFn = typing.Callable[[Node], None]
+"""A function that takes a Node and does something with it."""
+
+
+def visit(file: typing.TextIO, visit_fn: VisitFn) -> list[NodeParseError]:
+    """Parses an environment variable documentation file. The file must be
+    valid JSON and be a single object.
+
+    The file is completely parsed before any visit_fn is run. If there are
+    parsing errors, no visit_fn is run and the errors are returned.
+
+    If the file has no parsing errors, visit_fn is called on each node in
+    preorder traversal (visit node then its children) order.
+    """
+    file.seek(0)  # Ensure we are reading from the beginning of the file.
+
+    json_root = "file"
+    try:
+        docs = json.load(file)
+    except json.JSONDecodeError as e:
+        return [
+            NodeParseError(
+                json_root,
+                [ParseError(json_root, f"file is not valid JSON: {e}")], [])
+        ]
+    if not isinstance(docs, dict):
+        return [
+            NodeParseError(
+                json_root,
+                [ParseError(json_root, f"file is not a single JSON object")],
+                [])
+        ]
+
+    nodes_to_visit: list[Node] = []
+    parsing_errors: list[NodeParseError] = []
+
+    # The file is an implicit root node. Recursively descend into each top-level field.
+    for key, value in docs.items():
+        _recursively_parse(
+            1, json_root, key, value, nodes_to_visit, parsing_errors)
+
+    if len(parsing_errors) != 0:
+        return parsing_errors
+
+    for node in nodes_to_visit:
+        visit_fn(node)
+
+    return []
+
+
+def _recursively_parse(
+        level: int, parent_path: str, key: str, value: UnparsedJSON,
+        nodes_to_visit: list[Node], parsing_errors: list[NodeParseError]):
+    """Recursively visits a node and its children, in preorder traversal order (visit root then children)."""
+    path = _path(parent_path, key)
+
+    # Is this a group?
+    group, group_errors = _parse_group(path, value)
```

Comment:
The readability hit it due to wanting to return very detailed parsing errors.  If parsing fails, I want it be be obvious what field is missing/has a typo in the name.  So, we attempt to parse as a Group and get any errors from doing so.  If there were no errors, it is a Group and we process it.  Otherwise, we try to parse as a Variable.  If there were no errors, we process it.  If there were errors, then we could not parse the object as a Group nor a Variable.  We return the parsing errors from both attempts.  This makes it easier for callers of visit to print out detailed error messages.

For example, if we tried to parse:

```
{
  "My group": {
    "group_description": "A group",
    "members": {
       "MY_VAR": {
        "description": "A var"
        "typo": "string"
      }
    }
  }
}
```

```
Error at 'file.my group.MY_VAR':
  Could not parse as Group:
    'group_description' is a required field.
    'members' is a required field.
  Could not parse as a Variable:
    'type' is a required field.
    'typo' is not a valid field, valid fields are [...]
```

Hopefully from this error message it is easy to tell that the 'typo' field needs to change to 'type'

Reply:
PS: You could put those fields inside each of the classes that the list represents (group, variable, regex test etc)

Reply:
Do you think it would be easier to understand if the ensure_no_extra_fields call occurred after all the expected fields were parsed?